src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/StackOverflowClient.java

```diff
@@ -0,0 +1,143 @@
+package com.google.step.YOUR_PROJECT_NAME_HERE.external;
+
+import com.google.step.YOUR_PROJECT_NAME_HERE.data.Card;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.*;
+import java.util.regex.*;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.json.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public final class StackOverflowClient {
+  private static final String SEARCH_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s?"
+          + "order=desc&sort=activity&site=stackoverflow";
+  private static final String QUESTION_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s/answers?"
+          + "order=desc&sort=votes&site=stackoverflow";
+  // This url specify filter to generate answer body.
+  private static final String ANSWER_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/answers/%s?order"
+          + "=desc&sort=activity&site=stackoverflow&filter=!9_bDE(fI5";
+  private static final int ID_INDEX = 2;
+  private static final String ITEM_PARAMETER = "items";
+  private static final String TITLE_PARAMETER = "title";
+  private static final String BODY_PARAMETER = "body";
+  private static final String CODE_PARAMETER = "code";
+  private static final String ANSWER_ID_PARAMETER = "answer_id";
+  private static final int DESCRIPTION_LENGTH_PARAMETER = 200;
+
+  public Card search(String url) {
+    try {
+      Card card = getQuestion(url);
+      if (card == null) {
+        return null;
+      }
+      card = getAnswerId(card);
+      card = getAnswer(card);
+      return card;
+    } catch (URISyntaxException e) {
+      // Return null card if no valid card available.
+      return null;
+    }
+  }
+
+  /* Get the question id based on URL from the CSE result. */
+  private Card getQuestion(String url) throws URISyntaxException {
+    Card card = new Card();
+    URI uri = new URI(url);
+    // Parse the URL to get the question id.
+    String[] segments = uri.getPath().split("/");
+    String questionId = segments[ID_INDEX];
+    if (!Pattern.matches("[0-9]+", questionId)) {
+      return null;
+    }
+    String searchUrl = String.format(SEARCH_URL_TEMPLATE, questionId);
+    try {
+      JSONObject res = getResponse(searchUrl);
+      String title =
+          res.getJSONArray(ITEM_PARAMETER).getJSONObject(0).get(TITLE_PARAMETER).toString();
+      card.setLink(uri.toString());
+      // Store the id of the question in order to get the code body of the answer.
+      card.setCode(questionId);
```

Comment:
This is violating your data structure invariant, to put non-code in the code field.

Reply:
I know I said it was fine yesterday to pass partially-filled Cards around, but this is not what I meant. I would much rather the function return multiple values, as inelegant as it is, than reuse fields for things not advertised as such.

The core problem is that this class is not broken down into tasks in ways where the variables each task needs are in appropriate scopes. For example, here you're being forced to pass questionid and answerid within the Card to make it available to your other methods. Here's a different task breakdown that I believe solves the issue:
search() does the following:
1) get question id (in a private method if you want)
2) pass question id to a private method, get back the title
3) pass question id to a private method, get back the first answer id
4) pass answer id to a private method, get back answer body (I think this is HTML based on your code?)
5) pass answer body to a private method, get back description
6) pass answer body to a private method, get back code
7) finally create your Card object using title, uri, description, code

Thoughts?

(Note how each task either computes one piece of the Card, or does a network request to get a required parameter for a following task.)

(Side-effect: Card can be immutable after all, and all the getters and setters can be deleted.)

Reply:
To clarify: I'm okay with passing around partially-filled Cards if it makes it clearer how the code works (e.g. the paradigm is that each method fills out one or two fields of the Card, and this flow is clearly documented). I'm not okay with using the Card as hidden state - here, Card.code becomes an extra hidden argument to every method, and its meaning changes per method, which is _not_ clear at all.